api/base/filters.py

```diff
@@ -61,6 +64,163 @@ def filter_queryset(self, request, queryset, view):
         return queryset
 
 
+class MultiValueCharFilter(django_filters.BaseInFilter, django_filters.filters.CharFilter):
+
+    def filter(self, qs, value):
+        q = Q()
+        values = value or []
+        for value in values:
+            q |= Q(**{'{}__icontains'.format(self.name): value.strip()})
+        qs = qs.filter(q)
+
+        return qs
+
+class NullModelMultipleChoiceCaseInsensitiveField(forms.ModelMultipleChoiceField):
+
+    def clean(self, value):
+        # let a custom filter handle the actual filtering for null values later in the qs
+        if value == 'null':
+            return value
+
+        try:
+            return super(NullModelMultipleChoiceCaseInsensitiveField, self).clean(value)
+
+        except ValidationError as validation_error:
+            # Check to make sure the validation error wasn't because of a case sensitive relationship query
+            q = Q()
+            for choice in value:
+                q |= Q(**{'{}__iexact'.format(self.to_field_name): choice})
+            queryset = self.queryset.filter(q)
+
+            if not queryset:
+                raise validation_error
+
+            return queryset
+
+
+class NullModelMultipleChoiceFilter(django_filters.ModelChoiceFilter):
+
+    field_class = NullModelMultipleChoiceCaseInsensitiveField
+
+    def __init__(self, *args, **kwargs):
+        # use base_name to filter later on for a null relationship
+        self.base_name = kwargs.pop('base_name')
+        super(NullModelMultipleChoiceFilter, self).__init__(*args, **kwargs)
+
+    def filter(self, qs, value):
+        if value == 'null':
+            return qs.filter(Q(**{'{}__isnull'.format(self.base_name): True}))
+
+        q = Q()
+        values = value or []
+        for value in values:
+            q |= Q(**{self.name: value})
+        return qs.filter(q)
+
+
+class JSONAPIFilterSet(django_filters.rest_framework.FilterSet):
+
+    QUERY_PATTERN = re.compile(r'^filter\[(?P<fields>((?:,*\s*\w+)*))\](\[(?P<op>\w+)\])?$')
+    FILTER_FIELDS = re.compile(r'(?:,*\s*(\w+)+)')
+
+    # TODO - there must be a better way to recognize these
```

Comment:
You can do something like `if <class>._meta.get_field('<name>').many_to_many` to check for M2M. 
For date...
```
from django.db import connection

if 'timestamp' in <class>._meta.get_field('<name>').db_type(connection)
```
Not sure how hard it is to get the class in scope here, though

Reply:
Ah, that's good to know! I tried to get the class in scope here, but it's really tough, as far as I can tell my only args here are the field name and value being passed in by the filter....